Improved Documentation

ritikbanger opened the pull request:
<!-- Please describe your pull request here. -->

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- ~~Ensure you have provided tests - that demonstrates feature works or fixes the issue~~

This PR resolves the following issues- #396, #867, #802, #881 [comment](https://github.com/jenkinsci/docker/issues/881#issuecomment-882611061) ,#1267, #1468, #945.

Also, formatted the markdown file with proper indentations and spacings.

ritikbanger commented:
cc @timja

dduportal commented on README.md:
````diff
@@ -257,6 +264,50 @@ to indicate that this Jenkins installation is fully configured.
+To use Docker Compose with Jenkins, you can define a docker-compose.yml file that includes a Jenkins service and any other services that your Jenkins instance depends on, such as a database or cache. For example, the following docker-compose.yml file defines a Jenkins service that uses a MySQL database and a Redis cache:
+
+```
+version: "3"
````
```suggestion
```

version is deprecated in docker compose as per https://docs.docker.com/compose/compose-file/compose-versioning/

ritikbanger replied:
Addressed in [32ea849](https://github.com/jenkinsci/docker/pull/1530/commits/32ea849b7e5e4f7558632d19ee003537574095c7)

dduportal commented on README.md:
```diff
@@ -257,6 +264,50 @@ to indicate that this Jenkins installation is fully configured.
+
+### Docker Compose with Jenkins
+
+To use Docker Compose with Jenkins, you can define a docker-compose.yml file that includes a Jenkins service and any other services that your Jenkins instance depends on, such as a database or cache. For example, the following docker-compose.yml file defines a Jenkins service that uses a MySQL database and a Redis cache:
```
What is the purpose of using mysql or redis ? Since it is a « get started » example somhow, what do you think about using, instead of mysql and redis, an ssh-agent based on the jenkins/ssh-agent official image instead ?

ritikbanger replied:
make sense, addressed in https://github.com/jenkinsci/docker/pull/1530/commits/32ea849b7e5e4f7558632d19ee003537574095c7

dduportal commented on README.md:
```diff
@@ -257,6 +264,50 @@ to indicate that this Jenkins installation is fully configured.
+      - "8080:8080"
+    volumes:
+      - "./jenkins_home:/var/jenkins_home"
+    links:
```
Links should not be usef because it is deprecated : https://docs.docker.com/compose/compose-file/compose-file-v2/

dduportal commented on README.md:
```diff
@@ -257,6 +264,50 @@ to indicate that this Jenkins installation is fully configured.
+    ports:
+      - "8080:8080"
+    volumes:
+      - "./jenkins_home:/var/jenkins_home"
```
Bind mounting a volume from the host is not a good practise. What do you think about providing a named volume instead ?

it would avoid many beginners issues such as permissions issues, performances issues (if the bind mounted volume is shared through the VM like… Docker Desktop).

ritikbanger replied:
Added named volume. However, it is important to note that named volumes are not a replacement for proper backups. It is still a good idea to periodically backup your data to a secure location.

dduportal reviewed (changes requested):
Hello 👋 thanks for this huge documentation update !

I have added some comments : one major point is to avoid providing docker directives that are deprecated.

Also, we should not recommend bind mounting a volume: the sane default is to use docker named volume.

let me know if you want more pointers or details or help ( i got working examples)

ritikbanger commented:
> Hello 👋 thanks for this huge documentation update !
> 
> I have added some comments : one major point is to avoid providing docker directives that are deprecated.
> 
> Also, we should not recommend bind mounting a volume: the sane default is to use docker named volume.
> 
> let me know if you want more pointers or details or help ( i got working examples)

Thanks for the comments, i would be resolving them asap, would be glad if you add any working example to add in the doc.

dduportal commented:
> > Hello 👋 thanks for this huge documentation update !
> > I have added some comments : one major point is to avoid providing docker directives that are deprecated.
> > Also, we should not recommend bind mounting a volume: the sane default is to use docker named volume.
> > let me know if you want more pointers or details or help ( i got working examples)
> 
> Thanks for the comments, i would be resolving them asap, would be glad if you add any working example to add in the doc.

Hello @ritikbanger , you can find my "pet" project here: https://github.com/dduportal/jenkins-examples/tree/main/docker . The `docker-compose.yml` show a few services: you would be interested into `jenkins` (the controller) and the `jenkins-agent` (https://github.com/dduportal/jenkins-examples/blob/main/docker/docker-compose.yml#L15). As you can see, the `jenkins-agent` is clearly simple enough to fit the need of the quick start here.

Let me know if it is ok for you :)

lemeurherve commented on README.md:
```diff
@@ -257,6 +264,53 @@ to indicate that this Jenkins installation is fully configured.
+
+### Docker Compose with Jenkins
+
+To use Docker Compose with Jenkins, you can define a docker-compose.yml file that includes a Jenkins service and any other services that your Jenkins instance depends on, such as a database or ssh-agent. For example, the following docker-compose.yml file defines a Jenkins service that uses a Jenkins ssh-agent:
```
```suggestion
To use Docker Compose with Jenkins, you can define a docker-compose.yml file including a Jenkins instance and any other services it depends on, such as a database or an agent. For example, the following docker-compose.yml file defines a Jenkins controller and a Jenkins SSH agent:
```

lemeurherve reviewed (commented):
Thanks for the PR @ritikbanger!

Some suggestions below.

ritikbanger commented on README.md:
```diff
@@ -257,6 +264,55 @@ to indicate that this Jenkins installation is fully configured.
+
+The ssh-agent container is based on the jenkins/ssh-agent image and mounts a named volume ssh-agent-volume at /ssh-agent to persist the ssh keys.
+
+To start the Jenkins instance and the other services defined in the docker-compose.yml file, you can use the docker-compose up command. This will pull the necessary images from Docker Hub, if they are not already present on your system, and start the services in the background. You can then access the Jenkins web interface on port 8080 on your host system to configure and manage your Jenkins inst ...
```
make sense, addressed in https://github.com/jenkinsci/docker/pull/1530/commits/b1d964a5d8923c9a919b4432b40f66b618678e96

ritikbanger replied:
make sense, addressed in https://github.com/jenkinsci/docker/pull/1530/commits/b1d964a5d8923c9a919b4432b40f66b618678e96

lemeurherve commented on README.md:
````diff
@@ -257,6 +264,52 @@ to indicate that this Jenkins installation is fully configured.
+
+```
+
+This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent. The Jenkins container is based on the jenkins/jenkins:lts image and exposes the Jenkins web interface on port 8080.
+The jenkins_home volume is a named volume that is created and managed by Docker. It is mounted at /var/jenkins_home in the Jenkins container, and it will persist the Jenkins configuration and data.
+
+`jenkins_home` is a named volume created and managed by Docker. It is mounted at /var/jenkins_home in the Jenkins container, and will persist Jenkins configuration and data.
+
+Using a named volume can be a good alternative to a bind mount from the host if you want to store the data in a location managed by Docker and separated from the host file system. It can also be more portable as the data is stored within the Docker ecosystem and not tied to a specific host directory.
````
```suggestion
This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent.
The Jenkins container is based on the jenkins/jenkins:lts image and exposes the Jenkins web interface on port 8080.
The jenkins_home volume is a named volume that is created and managed by Docker.
It is mounted at /var/jenkins_home in the Jenkins container, and it will persist the Jenkins configuration and data.

`jenkins_home` is a named volume created and managed by Docker.
It is mounted at /var/jenkins_home in the Jenkins container, and will persist Jenkins configuration and data.

Using a named volume can be a good alternative to a bind mount from the host if you want to store the data in a location managed by Docker and separated from the host file system.
It can also be more portable as the data is stored within the Docker ecosystem and not tied to a specific host directory.
```

lemeurherve reviewed (commented):
Another suggestion: putting a sentence per line, for easier following change reviews.

dduportal commented on README.md:
````diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+To use Docker Compose with Jenkins, you can define a docker-compose.yml file including a Jenkins instance and any other services it depends on.
+For example, the following docker-compose.yml file defines a Jenkins controller and a Jenkins SSH agent:
+
+```
+version: '3.8'
+services:
+  jenkins:
+    image: jenkins/jenkins:lts
+    ports:
+      - "8080:8080"
+    volumes:
+      - jenkins_home:/var/jenkins_home
+  ssh-agent:
+    image: jenkins/ssh-agent
+    volumes:
+      - ssh-agent-volume:/ssh-agent
+volumes:
+  jenkins_home:
+  ssh-agent-volume:
+
+```
````
````suggestion
```yaml
services:
  jenkins:
    image: jenkins/jenkins:lts
    ports:
      - "8080:8080"
    volumes:
      - jenkins_home:/var/jenkins_home
  ssh-agent:
    image: jenkins/ssh-agent

volumes:
  jenkins_home:

```
````

Suggestion:
- Add explicit syntax color to the codeblock (use yaml)
- Remove the `version` attribute as it is deprecated (as per https://docs.docker.com/compose/compose-file/compose-versioning/#versioning)
- Remove the SSH agent volume: this snippet is focused on the Jenkins controller so better limiting the cognitive overload of the reader

dduportal commented on README.md:
````diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+
+```
+
+This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent.
````
```suggestion
This docker-compose.yml file creates two containers: one for Jenkins and one for the Jenkins SSH agent.
```

Use the same naming as used before the codeblock for coherency.

dduportal commented on README.md:
````diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+```
+
+This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent.
+The Jenkins container is based on the jenkins/jenkins:lts image and exposes the Jenkins web interface on port 8080.
````
```suggestion
The Jenkins container is based on the `jenkins/jenkins:lts` image and exposes the Jenkins web interface on port 8080.
```

Use "inline" font for filenames (implicit convention is markdown files)

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+
+This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent.
+The Jenkins container is based on the jenkins/jenkins:lts image and exposes the Jenkins web interface on port 8080.
+The jenkins_home volume is a named volume that is created and managed by Docker.
```
```suggestion
The `jenkins_home` volume is a [named volume](https://docs.docker.com/storage/volumes/) that is created and managed by Docker.
```

- Use "inline" font for code reference
- Add a link to Docker official documentation to provide actionnables to the reader

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+This docker-compose.yml file creates two containers: one for Jenkins and one for the ssh-agent.
+The Jenkins container is based on the jenkins/jenkins:lts image and exposes the Jenkins web interface on port 8080.
+The jenkins_home volume is a named volume that is created and managed by Docker.
+It is mounted at /var/jenkins_home in the Jenkins container, and it will persist the Jenkins configuration and data.
```
```suggestion
It is mounted at `/var/jenkins_home` in the Jenkins container, and it will persist the Jenkins configuration and data.
```

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+The jenkins_home volume is a named volume that is created and managed by Docker.
+It is mounted at /var/jenkins_home in the Jenkins container, and it will persist the Jenkins configuration and data.
+
+`jenkins_home` is a named volume created and managed by Docker.
+It is mounted at /var/jenkins_home in the Jenkins container, and will persist Jenkins configuration and data.
+
```
```suggestion
```

There is a duplication of these 2 lines with the above.

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+`jenkins_home` is a named volume created and managed by Docker.
+It is mounted at /var/jenkins_home in the Jenkins container, and will persist Jenkins configuration and data.
+
+Using a named volume can be a good alternative to a bind mount from the host if you want to store the data in a location managed by Docker and separated from the host file system.
+It can also be more portable as the data is stored within the Docker ecosystem and not tied to a specific host directory.
```
```suggestion
```

Limit the reader cognitive overload. This is an example snippet that should show a reference.

Bind mounting a an edge case that should not be advertised neither mentioned in this README (because not portable, not performant, not reliable, source of MANY problemsà.

Better to link to Docker volumes documentation instead (see my suggestion above).

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+Using a named volume can be a good alternative to a bind mount from the host if you want to store the data in a location managed by Docker and separated from the host file system.
+It can also be more portable as the data is stored within the Docker ecosystem and not tied to a specific host directory.
+
+The ssh-agent container is based on the jenkins/ssh-agent image and mounts a named volume `ssh-agent-volume` at /ssh-agent to persist the ssh keys.
```
```suggestion
The ssh-agent container is based on the jenkins/ssh-agent image and runs an SSH server to execute [Jenkins SSH Build Agent](https://plugins.jenkins.io/ssh-slaves/).
```

- Remove the volume mention
- Links to the Jenkins SSH Build Agent plugin to give actionnable to the reader

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+
+The ssh-agent container is based on the jenkins/ssh-agent image and mounts a named volume `ssh-agent-volume` at /ssh-agent to persist the ssh keys.
+
+To start the Jenkins instance and the other services defined in the docker-compose.yml file, you can use the `docker compose up` command.
```
```suggestion
To start the Jenkins instance and the other services defined in the `docker-compose.yml` file, run the `docker compose up -d`.
```

dduportal commented on README.md:
```diff
@@ -257,6 +264,59 @@ to indicate that this Jenkins installation is fully configured.
+
+To start the Jenkins instance and the other services defined in the docker-compose.yml file, you can use the `docker compose up` command.
+This will pull the necessary images from Docker Hub if they are not already present on your system, and start the services in the background.
+You can then access the Jenkins web interface on port 8080 on your host system to configure and manage your Jenkins instance using `docker compose up -d`.
```
```suggestion
You can then access the Jenkins web interface on `http://localhost:8080` on your host system to configure and manage your Jenkins instance (were `localhost` points to the published port by your Docker Engine).
```

dduportal commented on README.md:
```diff
@@ -276,6 +336,10 @@ Generally - you can copy it out - and then "docker pull" the image again - and y
 
 As always - please ensure that you know how to drive docker - especially volume handling!
 
+If you mount the Jenkins home directory to a directory on your Docker host up front, then the upgrade consists of `docker pull` and nothing more.
```
```suggestion
If you mount the Jenkins home directory to a [Docker named volume](https://docs.docker.com/storage/volumes/), then the upgrade consists of `docker pull` and nothing more.
```

Adds a link to docker volumes to the reader: it's up to the reader to decide wether or not to use a bind mount (which is discouraged by Docker unless you really know what you are doing), but we keep advertising a volume and that should still solve the upgrade problem (I can keep almost all your work here ❤️ )

dduportal reviewed (changes requested):
That is really a nice work!

I got a few suggestion:

- The "inline" code are nitpicks
- But the "volumes" releated suggestion are mandatory. I understand a lot of people are using a bind mount, but this should be considered a bad practise as per Docker official documentation.


(btw @ritikbanger if you are ok, I'm going to open an issue for documenting the "backup" topic to avoid making your work here waiting too much, considering we use named volumes).

ritikbanger commented:
> That is really a nice work!
> 
> I got a few suggestion:
> 
> * The "inline" code are nitpicks
> * But the "volumes" releated suggestion are mandatory. I understand a lot of people are using a bind mount, but this should be considered a bad practise as per Docker official documentation.
> 
> (btw @ritikbanger if you are ok, I'm going to open an issue for documenting the "backup" topic to avoid making your work here waiting too much, considering we use named volumes).

I have addressed all the comments, I think we are good to merge now. You can create another issue for the backup, any reference links would be helpful. You can create the issue and assign to me and I'll have another PR for it.

dduportal commented on README.md:
````diff
@@ -257,6 +264,53 @@ to indicate that this Jenkins installation is fully configured.
+For example, the following docker-compose.yml file defines a Jenkins controller and a Jenkins SSH agent:
+
+```yaml
+version: "3.8"
````
```suggestion
```

the keyword « version » is considered a legacy, and the version « 3.x » is only for docker swarm

ritikbanger replied:
Oops, missed this. Fixed in https://github.com/jenkinsci/docker/pull/1530/commits/739058c938316fad9ff45338dae2b079c846551e

dduportal reviewed (approved):
Nice job ! One last (non blocking suggestion)

dduportal reviewed (approved):
Thanks a lot for this contribution!
